### PR TITLE
ARROW-4975: [C++] Support concatenation of UnionArrays

### PR DESCRIPTION
**Pull request by mbrobbel**
This PR adds support for concatenation of union arrays. 

For sparse union arrays this is trivial: the type buffers and child arrays are concatenated like the other concatenate implementations.
For dense union arrays the following approach is used:
- The type buffers are concatenated.
- The child arrays are concatenated, but slicing is ignored. This makes building the offsets buffer more simple.
- For every input array the length of the different child arrays (concatenated up to that point) is tracked. By iterating over the types buffers these offsets can be applied to the values in the concatenated offsets buffer.

Does this make sense or should we slice child arrays (when required) and reflect this in the concatenated offsets buffer?

This PR also removes a check in `DenseUnionArray::Make` that rejected empty offsets buffers. This made it impossible to construct empty dense union arrays. I discussed removing this check with @bkietz.

### CODE REVIEW

File: cpp/src/arrow/array/concatenate_test.cc [hunk 1]
```diff
@@ -364,27 +364,89 @@ TEST_F(ConcatenateTest, DictionaryTypeNullSlots) {
...
+  auto offsets_sliced = ArrayFromJSON(int32(), "[1, 0, 1, 2, 3, 4, 3, 4, 5, 5]");
+  auto child_one_sliced =
+      ArrayFromJSON(boolean(), "[true, true, false, true, true, false]");
+  auto child_two_sliced = ArrayFromJSON(int8(), "[1, 2, 3, 1, 2, 3]");
+  ASSERT_OK_AND_ASSIGN(auto expected_sliced,
+                       DenseUnionArray::Make(*type_ids_sliced, *offsets_sliced,
+                                             {child_one_sliced, child_two_sliced}));
+  ASSERT_OK(expected_sliced->ValidateFull());
+  AssertArraysEqual(*expected_sliced, *concat_sliced_arrays);
+}
```
Comment (lidavidm) [inline cpp/src/arrow/array/concatenate_test.cc:448]: Can we also test concatenation of an array 1) which is not sliced, but whose children are sliced/have an offset? 2) which is sliced, whose children additionally have an offset?

**Review by lidavidm — APPROVED** [review]
Tricky indeed. I think just concatenating the dense union children (instead of trying to slice out only those elements which are referenced by an offset) is fine.

File: cpp/src/arrow/array/array_nested.cc [hunk 2]
```diff
@@ -708,10 +708,6 @@ DenseUnionArray::DenseUnionArray(std::shared_ptr<DataType> type, int64_t length,
 Result<std::shared_ptr<Array>> DenseUnionArray::Make(
     const Array& type_ids, const Array& value_offsets, ArrayVector children,
     std::vector<std::string> field_names, std::vector<type_code_t> type_codes) {
-  if (value_offsets.length() == 0) {
```
Comment (bkietz) [inline cpp/src/arrow/array/array_nested.cc:711]: :+1:

File: cpp/src/arrow/array/concatenate.cc [hunk 3]
```diff
@@ -349,7 +350,90 @@ class ConcatenateImpl {
...
+      for (size_t i = 0; i < in_.size(); i++) {
+        // Get sliced type ids and offsets.
+        auto type_ids = in_[i]->GetValues<int8_t>(1);
+        auto offset_values = in_[i]->GetValues<int32_t>(2);
+
+        // Iterate over all elements in the type buffer and append the updated
+        // offset to the concatenated offsets buffer.
+        for (auto j = 0; j < in_[i]->length; j++) {
+          int32_t offset;
+          if (internal::AddWithOverflow(offset_map[type_ids[j]], offset_values[j],
```
Comment (bkietz) [inline cpp/src/arrow/array/concatenate.cc:411]: Unfortunately, union type ids are slightly subtler than this: the values in the type ids buffer are drawn from UnionType::type_codes and are not necessarily a valid index into child_data. For example, consider the an array of `dense_union<int32=2, string=5>`. This array's type_ids buffer would be `[2, 5, 2]` to indicate int, string, int

UnionType::child_ids is a convenient mapping from type_id to child_data index

Comment (bkietz) [reply cpp/src/arrow/array/concatenate.cc:411]: Please rewrite the unit tests to use type_codes other than 0..n

File: cpp/src/arrow/array/concatenate_test.cc [hunk 4]
```diff
@@ -364,27 +364,89 @@ TEST_F(ConcatenateTest, DictionaryTypeNullSlots) {
...
+  auto type_ids_sliced = ArrayFromJSON(int8(), "[0, 1, 1, 0, 0, 0, 1, 1, 0, 1]");
+  auto offsets_sliced = ArrayFromJSON(int32(), "[1, 0, 1, 2, 3, 4, 3, 4, 5, 5]");
+  auto child_one_sliced =
+      ArrayFromJSON(boolean(), "[true, true, false, true, true, false]");
+  auto child_two_sliced = ArrayFromJSON(int8(), "[1, 2, 3, 1, 2, 3]");
+  ASSERT_OK_AND_ASSIGN(auto expected_sliced,
+                       DenseUnionArray::Make(*type_ids_sliced, *offsets_sliced,
+                                             {child_one_sliced, child_two_sliced}));
+  ASSERT_OK(expected_sliced->ValidateFull());
+  AssertArraysEqual(*expected_sliced, *concat_sliced_arrays);
```
Comment (bkietz) [inline cpp/src/arrow/array/concatenate_test.cc:447]: I think for for most of these arrays it's acceptable and more readable to come straight from JSON
```suggestion
  AssertArraysEqual(*ArrayFromJSON(dense_union({field("", boolean()), field("", int8())}), R"([
    [0, true],
    [1, 1],
    [1, 2],
    [0, true],
    [0, false],
    [0, true],
    [1, 3],
    [1, 1],
    [0, true],
    [1, 3]
  ])"), *concat_sliced_arrays);
```

The only case where you *need* to construct the dense union array with explicit children is when you're making an array with sliced children

**Review by bkietz — CHANGES_REQUESTED** [review]
This is looking great, thanks!

File: cpp/src/arrow/array/concatenate_test.cc [hunk 5]
```diff
@@ -364,27 +364,89 @@ TEST_F(ConcatenateTest, DictionaryTypeNullSlots) {
...
-  Check([this](int32_t size, double null_probability, std::shared_ptr<Array>* out) {
-    auto foo = this->GeneratePrimitive<Int8Type>(size, null_probability);
-    auto bar = this->GeneratePrimitive<DoubleType>(size, null_probability);
-    auto baz = this->GeneratePrimitive<BooleanType>(size, null_probability);
-    auto type_ids = rng_.Numeric<Int8Type>(size, 0, 2, null_probability);
-    auto value_offsets = rng_.Numeric<Int32Type>(size, 0, size, 0);
-    ASSERT_OK_AND_ASSIGN(
-        *out, DenseUnionArray::Make(*type_ids, *value_offsets, {foo, bar, baz}));
+  Check([this](int32_t size, double null_probabilities, std::shared_ptr<Array>* out) {
+    *out = rng_.ArrayOf(dense_union({
```
[the remark this answers is not in the document]
Comment (mbrobbel) [reply cpp/src/arrow/array/concatenate_test.cc:378]: The test already existed for both sparse and dense unions. However for the dense union it didn't generate valid arrays, so I updated that to use `ArrayOf`. I'll also update the sparse union generation to use `ArrayOf`.

**Review by bkietz — APPROVED** [review]
LGTM

CI failures seemed like unrelated flakes so I'm restarting those jobs to see if we can get to all-green

**Comment by bkietz** [comment]
merging

**Pull request closed without merging**